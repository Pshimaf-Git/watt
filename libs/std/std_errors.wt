/*
 Result type.
 | contains ok/err values, used
 | to work with errors and error propagation.
 | can contains ok or err at the same time.
 */
type Result(ok, err) {
    /*
     Result init,
     checks only one value is passed.
    */
    fn init {
        if self.ok != null and self.err != null {
            panic(
                'could not create result with (ok:'
                    + self.ok + ', err:' + self.err,
                'you can use only ok or err at the same time.'
            )
        }

        if self.ok == null and self.err == null {
            panic(
                'could not create result with both values are `null`',
                'if you\'re using `result.ok` or `result.err`, pass non-null value then.'
            )
        }

    }

    /*
     Returns true if Result is ok, else
     returns false
    */
    fn is_ok {
        return ok != null
    }

    /*
     Returns true if Result is err, else
     returns false
    */
    fn is_err {
        return err != null
    }

    /*
     Unwraps Result, if result is err,
     raises error, otherwise returns ok value
    */
    fn unwrap {
        if self.is_err() {
            panic(
                'unwrapping error result.',
                'error: ' + self.err
            )
        }
        else {
            return self.ok
        }
    }

    /*
     Unwraps Result, if result is ok,
     raises error, otherwise returns err value
    */
    fn unwrap_err {
        if self.is_ok() {
            panic(
                'unwrapping error, while result is ok.',
                'ok value: ' + self.ok
            )
        }
        else {
            return self.err
        }
    }

    /*
     Unwraps Result, if result is err,
     returns given value, otherwise
     returns ok value
    */
    fn unwrap_or(value) {
        if self.is_err() {
            return value
        }
        else {
            return self.ok
        }
    }

    /*
     Unwraps Result, if result is ok,
     returns given value, otherwise
     returns err value
    */
    fn unwrap_err_or(value) {
        if self.is_ok() {
            return value
        }
        else {
            return self.err
        }
    }
}

/*
 Result unit.
 | contains fast initializer
 | for Result type
*/
unit result {
    /*
     Result with ok value
    */
    fn ok(value) { return new Result(value, null) }

    /*
     Result with err value
    */
    fn err(err) { return new Result(null, err) }
}
