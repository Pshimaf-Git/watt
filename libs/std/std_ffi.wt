/*
 FFILibrary type
*/
type FFILibrary(raw) {
    fn call(name, args) {
        return ffi.__internal_call_fn(
            raw,
            name,
            args.internal
        )
    }
}

/*
 Library: ffi.
 | implementation of __internal_*
 | can be found in
 | > crate::vm::natives::libs::natives_ffi
*/
unit ffi {
    native __internal_load -> 'ffi@load'
    native __internal_load_fn -> 'ffi@load_fn'
    native __internal_call_fn -> 'ffi@call_fn'

    fn load(lib, functions) {
        library := __internal_load(lib)
        for function in functions.iter() {
            __internal_load_fn(
                library,
                function.get('name'),
                function.get('params').internal,
                function.get('out')
            )
        }
        return new FFILibrary(library)
    }
}