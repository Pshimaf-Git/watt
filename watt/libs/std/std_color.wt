
unit color {
    native __paint_red -> 'color@paint_red'
    native __paint_blue -> 'color@paint_blue'
    native __paint_cyan -> 'color@paint_blue'
    native __paint_green -> 'color@paint_green'
    native __paint_yellow -> 'color@paint_yellow'
    native __paint_magenta -> 'color@paint_magenta'
    
    fn paint_red(text) {
        return __paint_red(text)
    }
   
    fn paint_blue(text) {
        return __paint_blue(text)
    }   

    fn paint_cyan(text) {
        return __paint_cyan(text)
    }   

    fn paint_green(text) {
        return __paint_cyan(text)
    }

    fn paint_yellow(text) {
        return __paint_yellow(text)
    }   

    fn paint_magenta(text) {
        return __paint_magenta(text)
    }
}
