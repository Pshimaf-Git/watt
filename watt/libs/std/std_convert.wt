/*
 Library: convert.
 | implementation of to_int, to_float, to_long,
 | to_string, to_bool
 | can be found in:
 | > crate::vm::natives::libs::natives_convert
 */
unit convert {
    /*
     Converts value to int
    */
    native to_int -> 'convert@to_int'

    /*
     Converts value to float
    */
    native to_float -> 'convert@to_float'

    /*
     Converts value to string
    */
    native to_string -> 'convert@to_string'

    /*
     Converts value to bool
    */
    native to_bool -> 'convert@to_bool'
}