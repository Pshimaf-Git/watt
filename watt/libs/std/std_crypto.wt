/*
 Library: crypto. Unit: b64.
 | implementation of b64_encode, b64_decode
 | can be found in:
 | > crate::vm::natives::libs::natives_crypto
 */
unit b64 {
    /*
     To base64
    */
    native encode -> 'crypto@b64_encode'

    /*
     From base64
    */
    native decode -> 'crypto@b64_decode'
}

/*
 Library: crypto. Unit: sha256.
 | implementation of sha256_encode, sha256_decode
 | can be found in:
 | > crate::vm::natives::libs::natives_crypto
 */
unit sha256 {
    /*
     To sha256
    */
    native hash -> 'crypto@sha256'
}

/*
 Library: crypto. Unit: sha512.
 | implementation of sha512_encode, sha512_decode
 | can be found in:
 | > crate::vm::natives::libs::natives_crypto
 */
unit sha512 {
    /*
     To sha512
    */
    native hash -> 'crypto@sha512'
}

/*
 Library: crypto. Unit: sha224.
 | implementation of sha224_encode, sha224_decode
 | can be found in:
 | > crate::vm::natives::libs::natives_crypto
 */
unit sha224 {
    /*
     To sha224
    */
    native hash -> 'crypto@sha224'
}

/*
 Library: crypto. Unit: sha384.
 | implementation of sha384_encode, sha384_decode
 | can be found in:
 | > crate::vm::natives::libs::natives_crypto
 */
unit sha384 {
    /*
     To sha384
    */
    native hash -> 'crypto@sha384'
}

/*
 Library: crypto. Unit: md5.
 | implementation of md5
 | > crate::vm::natives::libs::natives_crypto
 */
unit md5 {
    /*
     To md5
    */
    native hash -> 'crypto@md5'
}