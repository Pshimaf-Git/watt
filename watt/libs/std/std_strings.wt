/*
 Library: Strings. String.
 | implementation of replace, strip, split,
 | trim, substring, char_at, replace_first,
 | replace_last, contains, index_of, last_index_of,
 | first_index_of
 | can be found in
 | > crate::vm::natives::libs::natives_strings
 */
type String(string) impl Display {

    /*
     Subs string from x to n
    */
    native __internal_substring -> 'strings@substring'

    /*
     Gives char at i. Has high runtime cost.
     Recommendation: store chars() if you want to use
     char_at a lot of times.
    */
    native __internal_char_at -> 'strings@char_at'

    /*
     Gets char list
    */
    native __internal_chars -> 'strings@chars'

    /*
     Splits string by delimiter
    */
    native __internal_split -> 'strings@split'

    /*
     Trims string
    */
    native __internal_trim -> 'strings@trim'

    /*
     Replace
    */
    native __internal_replace -> 'strings@replace'

    /*
     Replace n
    */
    native __internal_replace_n -> 'strings@replace_n'

    /*
     Replace range
    */
    native __internal_replace_range -> 'strings@replace_range'

    /*
     Contains
    */
    native __internal_contains -> 'strings@contains'

    /*
     Index of
    */
    native __internal_find -> 'strings@find'

    /*
     Last index of
    */
    native __internal_rfind -> 'strings@rfind'

    /*
     Push str to current string
    */
    native __internal_push -> 'strings@push'

    /*
     Length of string
    */
    native __internal_length -> 'strings@length'

    /*
     To lower case
    */
    native __internal_lower -> 'strings@lower'

    /*
     To upper case
    */
    native __internal_upper -> 'strings@upper'

    /*
     Native functions wrappers
    */
    fn replace(a, b) { return new String(self.__internal_replace(self.string, a, b)) }
    fn replace_n(a, b) { return new String(self.__internal_replace_n(self.string, a, b)) }
    fn replace_range(a, b) { return new String(self.__internal_replace_range(self.string, a, b)) }
    fn find(x) { return self.__internal_find(self.string, x) }
    fn rfind(x) { return self.__internal_rfind(self.string, x) }
    fn contains(x) { return self.__internal_contains(self.string, x) }
    fn trim { return new String(self.__internal_trim(self.string)) }
    fn split(x) {
        list := new List()
        list.internal = self.__internal_split(self.string, x)
        return list
    }
    fn sub(x, n) { return new String(self.__internal_substring(self.string, x, n)) }
    fn char_at(i) { return new String(self.__internal_char_at(self.string, i)) }
    fn chars {
        list := new List()
        list.internal = self.__internal_chars(self.string)
        return list
    }
    fn to_string {
        return string
    }
    fn push(str) {
        if typeof(str) == 'string' {
            self.__internal_push(self.string, str)
        } else {
            self.__internal_push(self.string, '' + str)
        }
        return self
    }
    fn length { return self.__internal_length(self.string) }
    fn to_lowercase { return new String(self.__internal_lower(self.string)) }
    fn to_uppercase { return new String(self.__internal_upper(self.string)) }
}

/*
 Library: Strings. Unit char.
 | implementation of is_digit, is_ascii_letter, as_int,
 | can be found in
 | > crate::vm::natives::libs::natives_strings
*/
unit char {
   // Checks char is digit with specific radix in range 2..36
   native is_digit -> 'char@is_digit'
   // Checks char is ascii letter
   native is_ascii_letter -> 'char@is_ascii_letter'
   // Returns char as int
   native as_int -> 'char@as_int'
}