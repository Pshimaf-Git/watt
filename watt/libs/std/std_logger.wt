import 'std.errors'
import 'std.color'
import 'std.time'
import 'std.io'

unit LoggerLevel {
    Trace := -15
    Debug := -10
    Info := -5
    Warn := 0
    Error := 5
    Fatal := 10
}

type FATAL(msg) impl Display {
    fn to_string() {
        return time.now().to_string() + color.paint_magenta(' FATAL: ') + msg
    }
}

type ERROR(msg) impl Display {
    fn to_string() {
        return time.now().to_string() + color.paint_red(' ERROR: ') + msg
    }
}

type WARN(msg) impl Display {
    fn to_string() {
        return time.now().to_string() + color.paint_yellow(' WARN: ') + msg
    }
}

type INFO(msg) impl Display {
    fn to_string() {
        return time.now().to_string() + color.paint_green(' INFO: ') + msg
    }
}

type DEBUG(msg) impl Display {
    fn to_string() {
        return time.now().to_string() + color.paint_blue(' DEBUG: ') + msg
    }
}

type TRACE(msg) impl Display {
    fn to_string() {
        return time.now().to_string() + color.paint_cyan(' TRACE: ') + msg
    }
}

type Logger(lvl) impl Display {
    fn to_string() {
        return 'logger with level ' + lvl
    }

    fn available(other_lvl) {
      return lvl <= other_lvl
    }

    fn Fatal(msg) {
        if available(LoggerLevel.Fatal) {
            io.println(new FATAL(msg))
        }
    }

    fn Error(msg) {
        if available(LoggerLevel.Error) {
            io.println(new ERROR(msg))
        }
    }

    fn Warn(msg) {
        if available(LoggerLevel.Warn) {
          io.println(new WARN(msg))
        }
    }

    fn Info(msg) {
        if available(LoggerLevel.Info) {
           io.println(new INFO(msg))
        }
    }

    fn Debug(msg) {
        if available(LoggerLevel.Debug) { 
            io.println(new DEBUG(msg))
        }
    }

    fn Trace(msg) {
        if available(LoggerLevel.Trace) {
            io.println(new TRACE(msg))
        }
    }
}
