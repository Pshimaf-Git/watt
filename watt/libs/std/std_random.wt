// imports
import 'std.math'
import 'std.time'

/*
 Random library.
 | implementation of: int, float, choose
 */
unit random {
    // Params (from C)
    a := 1664525
    c := 1013904223
    m := math.pow(2, 32)
    seed := time.now().mills()

    // Next seed
    fn next {
        self.seed = (self.a * self.seed + self.c) % self.m
        return self.seed
    }

    // Returns random float from 0 to 1
    fn _float {
        return next() / 4294967296
    }

    // Returns random int in range
    fn int(min, max) {
        return min + self.next() % (max - min + 1)
    }

    // Returns random float in range
    fn float(min, max) {
        return min + (_float() * (max-min))
    }
}
