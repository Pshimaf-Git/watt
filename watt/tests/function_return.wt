import 'std.io'

fn empty {}

fn value {
    return 12345678
}

fn value2 {
    return 'A string'
}

fn value3 {
    return value2() + ' but changed'
}

type Atom {}

fn value4 {
    return new Atom()
}

fn value5 {
    a := 1234
    b := 5678

    return a * b
}

io.println(empty())
io.println(value())
io.println(value2())
io.println(value3())
io.println(typeof(value4()))
io.println(value5())
