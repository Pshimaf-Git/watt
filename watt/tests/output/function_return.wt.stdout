Null
12345678
A string
A string but changed
Atom
7006652
